packages/api/src/OsdkObjectFrom.ts

```diff
@@ -137,44 +142,90 @@ type UnderlyingProps<
 
 export type IsNever<T> = [T] extends [never] ? true : false;
 
-type GetPropsKeys<
+type ExtractPropsKeysFromOldPropsStyle<
   Q extends ObjectOrInterfaceDefinition,
   P extends ValidOsdkPropParams<Q>,
 > = P extends "$all" ? PropertyKeys<Q>
   : Exclude<P, "$strict" | "$notStrict" | "$rid">;
 
-type GetProps<
+export type GetPropsKeys<
   Q extends ObjectOrInterfaceDefinition,
-  P extends ValidOsdkPropParams<Q>,
-> = P extends "$notStrict" ? CompileTimeMetadata<Q>["props"]
-  : CompileTimeMetadata<Q>["strictProps"];
+  P extends PropertyKeys<Q>,
+> = IsNever<P> extends true ? PropertyKeys<Q>
+  : IsAny<P> extends true ? PropertyKeys<Q>
+  : P;
+
+export type GetProps<
+  Q extends ObjectOrInterfaceDefinition,
+  OPTIONS extends never | "$notStrict" | "$rid" = never,
+> = IsNever<OPTIONS> extends true ? CompileTimeMetadata<Q>["strictProps"]
+  : (OPTIONS extends "$notStrict" ? CompileTimeMetadata<Q>["props"]
+    : CompileTimeMetadata<Q>["strictProps"]);
 
+/**
+ * @deprecated use `Osdk.Instance` or `YourType.OsdkObject`
+ */
 export type Osdk<
   Q extends ObjectOrInterfaceDefinition,
-  P extends ValidOsdkPropParams<Q> = "$all",
+  OPTIONS extends string = never,
+  P extends PropertyKeys<Q> = PropertyKeys<Q>,
 > =
-  & OsdkBase<Q>
-  & Pick<
-    GetProps<Q, P>,
-    GetPropsKeys<Q, P>
-  >
-  & {
-    readonly $link: Q extends { linksType?: any } ? Q["linksType"]
-      : Q extends ObjectTypeDefinition ? OsdkObjectLinksObject<Q>
-      : never;
-
-    readonly $as: <NEW_Q extends ValidToFrom<Q>>(type: NEW_Q | string) => Osdk<
-      NEW_Q,
-      ConvertProps<Q, NEW_Q, P>
+  // no middle options is simplest
+  IsNever<OPTIONS> extends true ? Osdk.Instance<Q, never, P>
+    // Options only includes the two allowed in the new style
+    : IsNever<Exclude<OPTIONS, "$notStrict" | "$rid">> extends true
+      ? Osdk.Instance<Q, OPTIONS & ("$notStrict" | "$rid"), P>
+    // else we are in the old style which was just Q and OPTIONS
+    // and OPTIONS was $things + prop names
+    : Osdk.Instance<
+      Q,
+      | ("$notStrict" extends OPTIONS ? "$notStrict" : never)
+      | ("$rid" extends OPTIONS ? "$rid" : never),
+      ExtractPropsKeysFromOldPropsStyle<Q, OPTIONS>
     >;
-  }
-  // We are hiding the $rid field if it wasn't requested as we want to discourage its use
-  & (IsNever<P> extends true ? {}
-    : string extends P ? {}
-    : "$rid" extends P ? { readonly $rid: string }
-    : {});
-
-export type OsdkObjectOrInterfaceFrom<
-  Q extends ObjectOrInterfaceDefinition,
-  P extends string = PropertyKeys<Q>,
-> = Osdk<Q, P>;
+
+export namespace Osdk {
```

Comment:
I'm not sure I fully understand the issues we were facing with the initial OSDK/why focusing on MyObject.Instance using what used to be OsdkObjectOrInterfaceFrom helps here

Reply:
`OsdkObjectOrInterfaceFrom` was a type we should have deleted. It was only used in the experimental subscription code, everything else was just `Osdk<>`.

The problem with keeping everything `Osdk<>` is that it leaves us in the world where we still have very complex types to evaluate what properties are available. Switching to this new type vastly simplifies the work the compiler was doing. However, we shipped 2.0 so we cannot JUST change `Osdk<>` because thats a breaking change. So this change makes it so that existing code that uses the `Osdk<>` object will keep working (at the continued expense of costly evaluation in the compiler) AND it makes it so that you can also use `Osdk<>` in the same shape as `Osdk.Instance<>` due to the back compat changes. This allows us in 3.x to later tell people to just use `Osdk<>` instead of `.Instance` if we so choose when we can completely delete the fallback behavior.

Because 2.0.0 has not gone to many people except internally and a few beta customers, this loss of `Osdk<>` is only going to be noticeable by those few people. Everyone else will just learn `Osdk.Instance`.